csharp/src/Ice/SlicSocket.cs

```diff
@@ -543,6 +512,62 @@ internal void TraceTransportFrame(string prefix, SlicDefinitions.FrameType type,
             Endpoint.Communicator.Logger.Trace(TraceLevels.TransportCategory, s.ToString());
         }
 
+        private protected override SocketStream CreateControlStream() =>
+            // We make sure to allocate the ID for the control stream right away. Otherwise, it would be considered
+            // like a regular stream and if serialization is enabled, it would acquire the semaphore.
+            new SlicStream(AllocateId(false), this);
+
+        private void ReadParameters(ReadOnlyMemory<byte> data, InputStream istr)
+        {
+            int dictionarySize = istr.ReadSize();
+            var parameters = new Dictionary<int, ReadOnlyMemory<byte>>(dictionarySize);
+            for (int i = 0; i < dictionarySize; ++i)
+            {
+                int key = istr.ReadVarInt();
+                int entrySize = istr.ReadSize();
+                if (!parameters.TryAdd(key, data.Slice(istr.Pos, entrySize)))
+                {
+                    throw new InvalidDataException($"duplicate Slic transport parameter `{key}");
+                }
+                istr.Skip(entrySize);
+            }
+
+            if (parameters.TryGetValue((int)ParameterKey.MaxBidirectionalStreams, out ReadOnlyMemory<byte> value))
+            {
+                BidirectionalStreamSemaphore = new AsyncSemaphore((int)value.Span.ReadVarULong().Value);
```

Comment:
Seems simpler to incorporate this unmarshaling into the earlier loop - this way,  no need for a temporary parameters dictionary.

Reply:
To me it's cleaner to separate the reading from the processing... but fine, I can change this.